src/adapter/src/coord/sequencer/inner/peek.rs

```diff
@@ -531,96 +525,104 @@ impl Coordinator {
 
+        let timestamp_context = determination.clone().timestamp_context;
         let stats = self
             .statistics_oracle(
                 ctx.session(),
                 &source_ids,
-                &determination.timestamp_context.antichain(),
+                &timestamp_context.antichain(),
                 true,
             )
             .await
             .unwrap_or_else(|_| Box::new(EmptyStatisticsOracle));
 
         mz_ore::task::spawn_blocking(
-            || "optimize peek (MIR)",
+            || "optimize peek",
             move || {
-                let pipeline = ||
-                 -> Result<
-                    Either<optimize::peek::GlobalMirPlan, optimize::copy_to::GlobalMirPlan>,
-                    AdapterError,
-                > {
-                    // In `explain_~` contexts, set the trace-derived dispatch
-                    // as default while optimizing.
-                    let _dispatch_guard = if let Some(explain_ctx) = explain_ctx.as_ref() {
-                        let dispatch = tracing::Dispatch::from(&explain_ctx.optimizer_trace);
-                        Some(tracing::dispatcher::set_default(&dispatch))
-                    } else {
-                        None
-                    };
-
-                    let _span_guard =
-                        tracing::debug_span!(target: "optimizer", "optimize").entered();
+                let pipeline = || -> Result<Either<optimize::peek::GlobalLirPlan, optimize::copy_to::GlobalLirPlan>, AdapterError> {
+                    let _dispatch_guard = explain_ctx.dispatch_guard();
 
                     let raw_expr = plan.source.clone();
 
-                    // Trace the pipeline input under `optimize/raw`.
-                    tracing::debug_span!(target: "optimizer", "raw").in_scope(|| {
-                        trace_plan(&raw_expr);
-                    });
-
                     match optimizer.as_mut() {
                         // Optimize SELECT statement.
                         Either::Left(optimizer) => {
-                            // HIR ⇒ MIR lowering and MIR ⇒ MIR optimization (local and global)
+                            // HIR ⇒ MIR lowering and MIR optimization (local)
                             let local_mir_plan = optimizer.catch_unwind_optimize(raw_expr)?;
-                            let local_mir_plan = local_mir_plan.resolve(ctx.session(), stats);
-                            let global_mir_plan =
-                                optimizer.catch_unwind_optimize(local_mir_plan)?;
+                            // Attach resolved context required to continue the pipeline.
+                            let local_mir_plan = local_mir_plan.resolve(timestamp_context, ctx.session(), stats);
+                            // MIR optimization (global), MIR ⇒ LIR lowering, and LIR optimization (global)
+                            let global_lir_plan = optimizer.catch_unwind_optimize(local_mir_plan)?;
```

Comment:
Is there still a need to have the optimizer pipeline split into two stages? Since the inputs to `resolve` are known before we start with the optimization, it seems like we could merge everything into one call.

Reply:
We could, but then probably the pipeline by constructing an ad-hoc struct with the `HirRelationExpr` and the associated context information.